lib/types/index.d.ts

```diff
@@ -1,98 +1,130 @@
 declare module 'node-dependency-injection' {
-    export class ContainerBuilder {
-        constructor ();
+    // Only for typings exports
 
-        addCompilerPass (compilerPass: any, ...args: any[]): void;
+    export type PassConfigHook = 'beforeOptimization'|'optimize'|'beforeRemoving'|'remove'|'afterRemoving';
 
-        compile (): void;
+    export type Parameter = string|boolean|object|any[];
 
-        findDefinition (key: any): any;
+    export type Argument = Reference|PackageReference|any;
 
-        findTaggedServiceIds (name: any): any;
+    export interface Extension {
+        load: Function;
+    }
 
-        get (id: any): any;
+    interface Logger {
+        warn(message?: any, ...optionalParams: any[]): void;
+    }
 
-        getDefinition (key: any): any;
+    export class InstanceManager {
+        constructor (containerBuilder: ContainerBuilder, definitions: Map<string, Definition>, alias: Map<string, string>);
 
-        getParameter (key: any): any;
+        getInstance (id: string, bypassPublic?: boolean): any;
 
-        has (key: any): any;
+        getInstanceFromDefinition (definition: Definition): any;
+    }
 
-        hasDefinition (key: any): any;
+    // Package exports
+    export class ContainerBuilder {
+        constructor (containerReferenceAsService?: boolean, defaultDir?: string|null);
 
-        hasParameter (key: any): any;
+        readonly defaultDir: string;
+        readonly containerReferenceAsService: boolean;
+        readonly definitions: Map<string, Definition>;
+        readonly instanceManager: InstanceManager;
+        readonly extensions: Extension[];
+        readonly services: Map<string, any>;
+        frozen: boolean;
+        logger: Logger;
 
-        isSet (id: any): any;
+        addCompilerPass (compilerPass: any, type?: PassConfigHook, priority?: number): void;
 
-        register (id: any, ...args: any[]): any;
+        compile (): void;
 
-        registerExtension (extension: any): void;
+        findDefinition (key: string): Definition;
 
-        remove (id: any): void;
+        findTaggedServiceIds (name: string): Map<any, any>;
 
-        removeDefinition (key: any): any;
+        get<T = any> (id: string): T;
 
-        set (id: any, instance: any): void;
+        getDefinition (key: string): Definition;
 
-        setAlias (alias: any, id: any): void;
+        getParameter<T extends Parameter> (key: string): T;
 
-        setDefinition (id: any, definition: any): any;
+        has (key: string): boolean;
 
-        setParameter (key: any, value: any): void;
+        hasDefinition (key: string): boolean;
 
-    }
+        hasParameter (key: string): boolean;
 
-    export class Definition {
-        constructor (...args: any[]);
+        isSet (id: string): boolean;
+
+        register (id: string, object: any, args: Argument[]): Definition;
+
+        registerExtension (extension: Extension): void;
```

Comment:
What are extensions? What do they? I didn't find any information in the documentation and couldn't understand it from the code.

Reply:
to be honest... no idea. I did that 3 years ago and I don't remember the real purpose of that feature.

Reply:
Ok then :)